test/xds/xds_client_ignore_resource_deletion_test.go

```diff
@@ -0,0 +1,473 @@
+// This helper creates a listener resource with using a custom listener created.
+// The test uses this listener to serve the gRPC server that is set up.
+func resourceWithListenerForGRPCServer(t *testing.T) (e2e.UpdateOptions, net.Listener) {
+	t.Helper()
+	lis, err := testutils.LocalTCPListener()
+	if err != nil {
+		t.Fatalf("testutils.LocalTCPListener() failed: %v", err)
+	}
+	t.Cleanup(func() { lis.Close() })
+
+	// Setup the management server to respond with the listener resources.
+	host, port, err := hostPortFromListener(lis)
+	if err != nil {
+		t.Fatalf("failed to retrieve host and port of server: %v", err)
+	}
+	listener := e2e.DefaultServerListener(host, port, e2e.SecurityLevelNone)
+	resources := e2e.UpdateOptions{
+		NodeID:    nodeID,
+		Listeners: []*listenerpb.Listener{listener},
+	}
+	return resources, lis
+}
+
+// This test creates a xds-enabled gRPC server with a listener resource and this
+// server features "ignore_resource_deletion". The test then verifies successful
+// RPCs to server. The test then removes the listener resource from the server
+// and verifies that server continues to server requests.
+func (s) TestListenerResourceDeletionOnServerIgnored(t *testing.T) {
+	mgmtServer := startManagementServer(t)
+	bs := generateBootstrapContents(t, mgmtServer.Address, true)
+	xdsR := xdsResolverBuilder(t, bs)
+	resources, lis := resourceWithListenerForGRPCServer(t)
+	updateCh, serve := setupGRPCServerWithModeChangeChannel(t, bs, lis)
+
+	go serve()
+
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	if err := mgmtServer.Update(ctx, resources); err != nil {
+		t.Fatal(err)
+	}
+
+	// Wait for the server to update to "serving" mode.
+	select {
+	case <-ctx.Done():
+		t.Fatal("Test timed out waiting for a mode change update.")
+	case mode := <-updateCh:
+		if mode != connectivity.ServingModeServing {
+			t.Fatalf("listener received new mode %v, want %v", mode, connectivity.ServingModeServing)
+		}
+	}
+
+	// Create a ClientConn and make a successful RPCs.
+	cc, err := grpc.Dial(lis.Addr().String(), grpc.WithTransportCredentials(insecure.NewCredentials()), grpc.WithResolvers(xdsR))
+	if err != nil {
+		t.Fatalf("failed to dial local test server: %v", err)
+	}
+	defer cc.Close()
+
+	if err := verifyRPCtoAllEndpoints(cc); err != nil {
+		t.Fatal(err)
+	}
+
+	// Update without a listener resource.
+	if err := mgmtServer.Update(ctx, e2e.UpdateOptions{
+		NodeID:    nodeID,
+		Listeners: []*listenerpb.Listener{},
+	}); err != nil {
+		t.Fatal(err)
+	}
+
+	// Make RPCs every 100 mgmtServer for 1s and verify that the serving mode does
+	// not change on server.
+	timer := time.NewTimer(500 * time.Millisecond)
+	ticker := time.NewTicker(50 * time.Millisecond)
+	t.Cleanup(ticker.Stop)
+	for {
+		if err := verifyRPCtoAllEndpoints(cc); err != nil {
+			t.Fatal(err)
+		}
+		select {
+		case <-timer.C:
+			return
+		case mode := <-updateCh:
+			t.Fatalf("Listener received new mode: %v", mode)
```

Comment:
Nit: `t.Fatalf("Server switched to mode: %v when no switch was expected", mode)

Reply:
done